certification/internal/policy/operator/has_v2_related_image_schema_version.go

```diff
@@ -0,0 +1,182 @@
+package operator
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/operator-framework/api/pkg/manifests"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// RelatedImagesAreSchemaVersion2Check is part of the Operator policy and implements
+// the Check interface.
+type RelatedImagesAreSchemaVersion2Check struct{}
+
+// Validate checks the image manifest for each related image referenced in a
+// ClusterServiceVersion and ensures that the schema version used is version 2.
+func (p *RelatedImagesAreSchemaVersion2Check) Validate(imgRef certification.ImageReference) (bool, error) {
+	imageToSchemaVersion, err := p.getDataToValidate(imgRef)
+	if err != nil {
+		return false, fmt.Errorf("%w: %s", errors.ErrRunContainerFailed, err)
+	}
+
+	return p.validate(imageToSchemaVersion)
+}
+
+// getDataToValidate pulls a ClusterServiceVersion from the input operator bundle,
+// checks the ClusterServiceVersion's related images declaration, and assembles a list of
+// images and their image manifest schema version values.
+func (p *RelatedImagesAreSchemaVersion2Check) getDataToValidate(imgRef certification.ImageReference) (map[string]int, error) {
+	manifestsDir := path.Join(imgRef.ImageFSPath, "manifests")
+	csv, err := p.readBundle(manifestsDir)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	relatedImages, err := p.getRelatedImagesForCSV(csv)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	imageSchemaVersionMap, err := p.inspectSchemaVersionForImage(relatedImages)
+	if err != nil {
+		log.Error(err)
+		return nil, fmt.Errorf("%w: %s", errors.ErrDeterminingRelatedImageSchemaVers, err)
+	}
+
+	return imageSchemaVersionMap, nil
+}
+
+// readBundle will accept the manifests directory where a bundle is expected to live,
+// and walks the directory to find all bundle assets.
+func (p *RelatedImagesAreSchemaVersion2Check) readBundle(manifestsDir string) (*operatorsv1alpha1.ClusterServiceVersion, error) {
```

Comment:
Can we have this method in util/container instead? It seems a couple of checks could leverage this method.

Reply:
Great idea, I've made the function call generic and am leveraging it in the new policy definition.

A note: I did not change the way this works in the old engine's equivalent policy definition as we're moving away from it anyway.